core/trino-main/src/main/java/io/trino/operator/window/pattern/FeedablePagesWindowIndex.java

```diff
@@ -0,0 +1,239 @@
+ * The purpose of this class is to provide input for aggregations
+ * in a consistent way, regardless of whether the aggregation
+ * uses an input channel, or a runtime-evaluated expression as
+ * the argument.
+ * <p>
+ * The latter case applies to aggregate functions within the
+ * DEFINE and MEASURES clauses of row pattern recognition.
+ * The aggregated argument can be based on CLASSIFIER() or
+ * MATCH_NUMBER() results which are not present in the input,
+ * and thus needs to be evaluated at runtime while processing
+ * a row.
+ * <p>
+ * E.g. for the aggregate function array_agg(lower(CLASSIFIER(X))),
+ * there is no input channel containing the aggregated argument,
+ * so it is evaluated row by row at runtime and passed in an additional
+ * "channel" available to the aggregation's accumulator by a standard
+ * WindowIndex interface.
+ */
+public class FeedablePagesWindowIndex
+        implements WindowIndex
+{
+    private final PagesIndex pagesIndex;
+    private final int start;
+    private final int size;
+    private final List<Block> feedableChannels;
+    private final List<Type> feedableTypes;
+    private int fedPosition;
+    private final List<ArgumentComputation> feeders;
+    private final int firstFeedableChannel;
+
+    // View of the underlying PagesIndex used by feeders. It does not contain feedable channels.
+    private final WindowIndex windowIndex;
+
+    public FeedablePagesWindowIndex(PagesIndex pagesIndex, int start, int end, List<ArgumentComputation> feeders)
+    {
+        requireNonNull(pagesIndex, "pagesIndex is null");
+        checkPositionIndex(start, pagesIndex.getPositionCount(), "start");
+        checkPositionIndex(end, pagesIndex.getPositionCount(), "end");
+        checkArgument(start < end, "start must be before end");
+        requireNonNull(feeders, "feeders is null");
+
+        this.pagesIndex = pagesIndex;
+        this.start = start;
+        this.size = end - start;
+        this.feeders = ImmutableList.copyOf(feeders);
+        this.feedableTypes = feeders.stream()
+                .map(ArgumentComputation::getOutputType)
+                .collect(toImmutableList());
+        this.feedableChannels = new ArrayList<>();
+        for (int i = 0; i < feedableTypes.size(); i++) {
+            feedableChannels.add(null);
+        }
+        this.fedPosition = -1;
+        this.firstFeedableChannel = pagesIndex.getTypes().size();
+        this.windowIndex = new PagesWindowIndex(pagesIndex, start, end);
+    }
+
+    /**
+     * Compute values in feedable channels which are in the `argumentChannels` list.
+     * Put `UNIMPLEMENTED_BLOCK` in other feedable channels.
+     *
+     * @param position position relative to partition start
+     */
+    public void feedRow(int position, List<Integer> argumentChannels, int label, long matchNumber)
+    {
+        fedPosition = position;
+        for (int i = 0; i < feedableTypes.size(); i++) {
+            feedableChannels.set(i, UNIMPLEMENTED_BLOCK);
```

Comment:
What's the point of this? Will anyone ever try to reference those blocks, in which case they'll get an error? If so, maybe just leave these as nulls and get rid of UNIMPLEMENTED_BLOCK.